compiler/src/dmd/lexer.d

```diff
@@ -701,12 +704,102 @@ class Lexer
                     }
 
-                    Identifier id = Identifier.idPool((cast(char*)t.ptr)[0 .. p - t.ptr], false);
+                    Identifier id;
+
+                    if (universalCharacterName4 == 0 && universalCharacterName8 == 0)
+                    {
+                        id = Identifier.idPool((cast(char*)t.ptr)[0 .. p - t.ptr], false);
+                    }
+                    else
+                    {
+                        auto priorValidation = t.ptr[0 .. p - t.ptr];
+                        const(char)* priorVPtr = priorValidation.ptr;
+                        const possibleLength = (
+                            priorValidation.length - (
+                                    (universalCharacterName4 * 6) +
+                                    (universalCharacterName8 * 10)
+                            )) + (
+                                (universalCharacterName4 * 3) +
+                                (universalCharacterName8 * 4)
+                        );
+
+                        char[64] buffer = void;
+                        SmallBuffer!char sb = SmallBuffer!char(possibleLength + 1, buffer[]);
+
+                        char[] storage = sb.extent;
+                        size_t offset;
+
+                        while(priorVPtr < &priorValidation[$-1])
+                        {
+                            if (*priorVPtr == '\\')
+                            {
+                                dchar tempDchar = 0;
+                                uint times;
+
+                                // universal character name (C)
+                                if (priorVPtr[1] == 'u')
+                                    times = 4;
+                                else if (priorVPtr[1] == 'U')
+                                    times = 8;
+                                else
+                                    assert(0, "ICE: Universal character name is 2 or 4 bytes only");
+                                priorVPtr += 2;
+
+                                foreach(i; 0 .. times)
+                                {
+                                    char c = priorVPtr[i];
+
+                                    if (c >= '0' && c <= '9')
+                                        c -= '0';
+                                    else if (c >= 'a' && c <= 'f')
+                                        c -= 'a' - 10;
+                                    else if (c >= 'A' && c <= 'F')
+                                        c -= 'A' - 10;
+
+                                    tempDchar <<= 4;
+                                    tempDchar |= c;
+                                }
+
+                                priorVPtr += times - 1;
+
+                                if (!utf_isValidDchar(tempDchar))
+                                {
+                                    warning(t.loc, "char 0x%x code point is outside of valid UTF-8 encoding range and cannot be in an identifier", tempDchar);
```

Comment:
"shall not" -> error
```suggestion
                                    error(t.loc, "character 0x%x is not allowed in an identifier", tempDchar);
```

Although, given a few random characters a test, gcc and clang refuse characters that are within the range that isn't explicitly rejected in spec.

i.e:
```
int \uE0000(void)
{
    return 0;
}
```
Maybe its rejected in ISO/IEC 10646, but I don't have access to that spec paper?

Reply:
I tried ``int \U000E0000;`` with clang and gcc, same results as above.
Note: I'm explicitly setting ``-std=c23``

I got the previous value from looking at the character table for C23 and grabbing something outside its range.